[FLINK-26261] Wait for JobManager deployment ready before accessing REST API

tweise opened the pull request:
First part following resource dependencies in the reconciler logic: Wait for the JM deployment to be ready before trying to access the JM. This will be built upon for improving the error detection/handling later on.

tweise commented:
@gyfora PTAL

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:
```diff
@@ -109,6 +119,22 @@ public DeleteControl cleanup(FlinkDeployment flinkApp, Context context) {
+                                        "JobManager deployment {} in namespace {} is ready",
+                                        flinkApp.getMetadata().getName(),
+                                        flinkApp.getMetadata().getNamespace());
+                                jobManagerDeployments.put(flinkApp, deployment);
```
Shouldn't we check and put into the jobManagerDeployments befre we try to observe? It seems now we introduce an extra refesh between every reconcile and observe.

tweise replied:
We do check before we observe. The deployment check is executed once, it does not introduce an extra refresh. It also works when the operator is restarted and finds existing deployments.

gyfora replied:
Oh I think I see what you mean. Do you mean that we check after the initial reconcolitiation? (right after the job was deployed).

My suggestion was to move before line 108 just to be safe, but maybe it's not necessary

tweise replied:
The current implementation checks job status only after the deployment was observed. I think my earlier reply was misleading.

tweise replied:
Note that that there is other logic in `JobStatusObserver` that will exit when not ready to watch, confusingly returning true. I did not want to touch that in the PR. My goal is to implement the terminal error check when the deployment never shows up, for example due to invalid service account.

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:
```diff
@@ -88,17 +92,23 @@ public DeleteControl cleanup(FlinkDeployment flinkApp, Context context) {
                 operatorNamespace,
                 kubernetesClient,
                 true);
+        jobManagerDeployments.remove(flinkApp);
```
Should we also remove from this map when we cancel the flink job in the Flink service?

gyfora replied:
Otherwise after an upgrade the observer might hit the same issue

tweise replied:
Good catch, will look into that tomorrow.

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:
```diff
@@ -109,6 +119,22 @@ public DeleteControl cleanup(FlinkDeployment flinkApp, Context context) {
             throw new ReconciliationException(e);
         }
+
+        if (!jobManagerDeployments.containsKey(flinkApp)) {
```
Maybe we could extract the checking and loading of the deployment from the secondaryResource into a new method in the controller to keep the main reconcile flow as simple as possible

morhidi replied:
We could also move the whole logic into JobStatusObserver passing the context object.

tweise replied:
It would scatter the resource watching logic. I would prefer to keep it here since it needs to work in tandem with `prepareEventSources`

gyfora replied:
I also agree with @morhidi that this probably fits best into the observer itself because that is the component responsible for checking the running jobs already. The observer could also prepare the event sources that we would just pass in the controller.

tweise replied:
Refactored the logic and `JobStatusObserver` can be removed / consolidated with JobReconciler in a follow-up.

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:
```diff
@@ -144,7 +170,16 @@ private void updateForReconciliationError(FlinkDeployment flinkApp, String err)
         //        return List.of(new PerResourcePollingEventSource<>(
         //                new FlinkResourceSupplier, context.getPrimaryCache(), POLL_PERIOD,
         //                FlinkApplication.class));
```
I think we can remove these comments now that we have some logic here :)

gyfora commented:
I wonder if caching the deployment is even necessary, it adds complexity and can lead to cases where the deployment disappears and we do not notice. The reconcile loops should not run too often for this to stress kubernetes, what do you think @tweise ?

morhidi commented:
I tested your PR locally and hit this error, when deleting the CR @tweise did you notice?

```
2022-02-24 11:44:24,277 o.a.f.k.o.c.FlinkDeploymentController [INFO ] [default.basic-example] Cleaning up application cluster basic-example
Exception in thread "pool-3-thread-4" java.lang.IllegalStateException: Cached custom resource must be present at this point
	at io.javaoperatorsdk.operator.processing.event.EventProcessor.lambda$isCacheReadyForInstantReconciliation$1(EventProcessor.java:248)
```

morhidi commented:
> I wonder if caching the deployment is even necessary, it adds complexity and can lead to cases where the deployment disappears and we do not notice. The reconcile loops should not run too often for this to stress kubernetes, what do you think @tweise ?

I agree with @gyfora here. The refresh loop is 1 min at the moment, shouldn't be a big deal.

tweise commented:
> I wonder if caching the deployment is even necessary, it adds complexity and can lead to cases where the deployment disappears and we do not notice. The reconcile loops should not run too often for this to stress kubernetes, what do you think @tweise ?

We actually don't need to cache the whole objects for this purpose. However, I would like to limit the checking for secondary resources unless really necessary as that may become a scalability concern. How frequently we check the job status should ideally be independent from this logic.

tweise commented:
> I tested your PR locally and hit this error, when deleting the CR @tweise did you notice?
> 
> ```
> 2022-02-24 11:44:24,277 o.a.f.k.o.c.FlinkDeploymentController [INFO ] [default.basic-example] Cleaning up application cluster basic-example
> Exception in thread "pool-3-thread-4" java.lang.IllegalStateException: Cached custom resource must be present at this point
> 	at io.javaoperatorsdk.operator.processing.event.EventProcessor.lambda$isCacheReadyForInstantReconciliation$1(EventProcessor.java:248)
> ```

I also see it now. It is related to the deployment event watching as it goes away when disabling `prepareEventSources`

tweise commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:
```diff
@@ -109,6 +118,19 @@ public DeleteControl cleanup(FlinkDeployment flinkApp, Context context) {
+                        flinkApp.getMetadata().getNamespace());
+                jobManagerDeployments.add(flinkApp.getMetadata().getSelfLink());
+                // reschedule for immediate job status check
+                return UpdateControl.updateStatus(flinkApp).rescheduleAfter(0);
```
We will still need a delay here as even after the pod is ready it still takes a bit for the REST server to listen at the port. This is how I currently see the job status check still hitting a timeout. BTW @gyfora job status checks blocking the operator threads isn't so nice!

morhidi replied:
Try @ControllerConfiguration(generationAwareEventProcessing = true)

morhidi replied:
In this case, the reconcile loop will kick off after 60 seconds. Can be a simple workaround until we find a solution for this

tweise replied:
The annotation does not seem to change anything wrt the original error. There is a separate reschedule interval for this condition, which in absent more sophisticated readiness check I have set to 5s for now - that seems to avoids the timeout exception.

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:
```diff
@@ -109,6 +118,19 @@ public DeleteControl cleanup(FlinkDeployment flinkApp, Context context) {
             throw new ReconciliationException(e);
         }
+
+        if (!jobManagerDeployments.contains(flinkApp.getMetadata().getSelfLink())) {
+            Optional<Deployment> deployment = context.getSecondaryResource(Deployment.class);
+            if (deployment.isPresent()) {
+                LOG.info(
+                        "JobManager deployment {} in namespace {} is ready",
+                        flinkApp.getMetadata().getName(),
+                        flinkApp.getMetadata().getNamespace());
+                jobManagerDeployments.add(flinkApp.getMetadata().getSelfLink());
+                // reschedule for immediate job status check
+                return UpdateControl.updateStatus(flinkApp).rescheduleAfter(0);
```
could we move this whole thing into a `boolean checkAndCacheDeployment(...)` -or something similar- method?

tweise replied:
I moved it into into separate method for now. Will see how it can be arranged better when I have the error/timeout check in place.

gyfora commented:
@wangyang0918 if you have some time to look at this we could use a bit of extra feedback :)

wangyang0918 commented:
I am getting to this PR now.

wangyang0918 commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:
```diff
@@ -145,11 +190,16 @@ private void updateForReconciliationError(FlinkDeployment flinkApp, String err)
+                kubernetesClient
+                        .apps()
+                        .deployments()
+                        .inAnyNamespace()
```
Maybe we do not need to watch all the namespaces if `FLINK_OPERATOR_WATCH_NAMESPACES` configured.

tweise replied:
I would prefer that. Need to figure out how to get the controller config from the context and set multiple namespaces.

wangyang0918 commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:
```diff
@@ -145,11 +190,16 @@ private void updateForReconciliationError(FlinkDeployment flinkApp, String err)
+                        .apps()
+                        .deployments()
+                        .inAnyNamespace()
+                        .withLabel("type", "flink-native-kubernetes")
```
Using constants in `org.apache.flink.kubernetes.utils.Constants` could make us aware of upstream changes.

wangyang0918 commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:
```diff
@@ -114,6 +126,39 @@ public DeleteControl cleanup(FlinkDeployment flinkApp, Context context) {
+                            "JobManager deployment {} in namespace {} is ready",
+                            flinkApp.getMetadata().getName(),
+                            flinkApp.getMetadata().getNamespace());
+                    jobManagerDeployments.add(flinkApp.getMetadata().getSelfLink());
```
Do we need to remove the current flinkApp from cache `jobManagerDeployments` when the replicas is not enough? For example, the JobManager crashed for a while.

tweise replied:
Maybe, the job status check would fail in that case and we could handle it there. We can address this as follow-up. There is a lot of work that needs to go into the reconciliation logic based on what I noticed while working on this PR.

wangyang0918 reviewed (commented):
I think this PR is a good step to avoid unnecessary reconciliation. But it seems that I still find some `java.util.concurrent.TimeoutException` when accessing the rest service even after `PORT_READY_DELAY_SECONDS`. Maybe my minikube is a little slow or JobManager does not have enough resources(0.5 cpu).

tweise commented:
@gyfora this is working pretty well now, I also added a port readiness check. But the current implementation with control logic scattered over job status observer and reconcilers makes it pretty hard to implement this in a cleaner way. I'm working on some refactoring and hopefully that will lead to simplification. Hope to have it done soon to avoid larger merge conflicts.

gyfora commented:
Sounds good @tweise !

tweise commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java:
```diff
@@ -114,6 +126,39 @@ public DeleteControl cleanup(FlinkDeployment flinkApp, Context context) {
+
+    private UpdateControl<FlinkDeployment> checkDeployment(
+            FlinkDeployment flinkApp, Context context) {
+        if (!jobManagerDeployments.contains(flinkApp.getMetadata().getSelfLink())) {
```
It is a unique reference that avoids caching entire resource objects and duplicate key construction. Why should the field not exist? Did you encounter it somewhere?

wangyang0918 replied:
I do not find the `selfLink` in my minikube with version `v1.23.0`. Could we simply use the `metadata.name` here?

![image](https://user-images.githubusercontent.com/15904523/155877551-7bc72a22-5c4d-4130-a3ad-04e03aa26b43.png)

tweise replied:
I see, it appears that selflink was deprecated. We need a unique identifier across namespaces. `UID` should do the trick.

tweise commented:
@gyfora I think after the refactor this looks much better now. Minor improvements can be dealt with as follow-up. I think there will be significant work necessary to make the reconciliation more robust. I'm wondering if it should be implemented as a state machine. The job upgrade part needs to be looked at. What happens when the previous job never came up due to a code issue and user pushes an update to fix the problem. This should be possible and we need to have test coverage for it also.

tweise commented:
> I think this PR is a good step to avoid unnecessary reconciliation. But it seems that I still find some `java.util.concurrent.TimeoutException` when accessing the rest service even after `PORT_READY_DELAY_SECONDS`. Maybe my minikube is a little slow or JobManager does not have enough resources(0.5 cpu).

I changed the delay to 10s and added a port readiness check. The 10s may still not be enough in some situations for the REST service to get ready (after the port is ready). That is covered by the retry though.

gyfora commented:
I will have to take a look again after the refactor, but there is also in the ticket opened in the meantime: https://issues.apache.org/jira/browse/FLINK-26377

cc @Aitozi
